src/install/dracut-install.c

```diff
@@ -539,26 +538,26 @@ static int resolve_deps(const char *src)
                 }
         }
 
-        /* run ldd */
-        _asprintf(&cmd, "%s %s 2>&1", ldd, fullsrcpath);
+        int fds[2];
+        if (pipe2(fds, O_CLOEXEC) == -1 || (fptr = fdopen(fds[0], "r")) == NULL) {
+                log_error("Error '%s' initiating pipe stream for %s", strerror(errno), ldd);
+                exit(EXIT_FAILURE);
+        }
 
-        log_debug("%s", cmd);
+        log_debug("%s %s", ldd, fullsrcpath);
+        if (vfork() == 0) {
+                dup2(fds[1], 1);
+                dup2(fds[1], 2);
+                execlp(ldd, ldd, fullsrcpath, (char *)NULL);
```

Comment:
This will break at least the "command not found" case, that will be considered successful (`ret = 0`). However current `popen()` code is also broken for non-english locales: `sh` and `ldd` output are internationalized and error messages do not match.

Reply:
Wait, this actually depends on the "command not found" bash output? `sh -c` obviously outputs `sh: 1: ldd: not found`, like it has for the past 30+ years. Wild. Added a "command not found" write to compensate.

The latter fixed by setting `LC_ALL=C`.

Reply:
I wonder if checking *wstatus* could simplify things, like:

```c
        pid_t pid = fork();
        if (pid == 0) {
                dup2(fds[1], 1);
                dup2(fds[1], 2);
                putenv("LC_ALL=C");
                execlp(ldd, ldd, fullsrcpath, (char *)NULL);
                _exit(EXIT_FAILURE);
        }
        close(fds[1]);

        while (waitpid(pid, &ret, 0) < 0) {
                if (errno != EINTR) {
                        ret = 1;
                        break;
                }
        }

        if (ret != 0) {
                log_error("ERROR: failed to run '%s %s'", ldd, fullsrcpath);
                return 1;
        }

        while (getline(&buf, &linesize, fptr) >= 0) {

        ...
```

(dropped `err`)

Then clean up `strstr()` calls below, removing the ones which correspond to unclean exit codes.

Reply:
well, there's three strings
> "you do not have execution permission"
> "unable to find sysroot"
> "command not found"

The last one is obviously from bash, but (my) ldd (or ld.so) doesn't contain the other two. Are they all foldable together to $? != 0?

Reply:
I am on Fedora 36 here btw. "you do not have execution permission" was added in

https://github.com/dracutdevs/dracut/commit/b4dc22cab7b8b071c8a150b9c61edaa247bacb6a

hence it must stay too.

Reply:
**Or**, if the `noexec` case makes `ldd` exit code != 0, then drop it. Needs testing.

Reply:
Turns out this is a Debian patch, `all/local-ldd.diff`: 
```
-    test -x "$file" || echo 'ldd:' $"\
-warning: you do not have execution permission for" "\`$file'" >&2
```

Applied != 0 => ret += 1; retained all except for "command not found"